### DISCUSSION TITLE
2.x 重复打包问题汇总

### DISCUSSION THREAD

**Discussion by janryWang** [alibaba/formily#1406] [Ideas]
这个问题已经有很多人提过了，记录一下，为了保证所有用户安装依赖数量不要太多，所以在@formily内部包之间的依赖关系全部走的dependencies，而非peerDependencies，同时dependencies版本号全部写死，这样也方便回滚，但是带来的问题就是：
- webpack 关闭系统软链接(systemlinks)一定会导致重复打包
- 项目里依赖的@formily/core @formily/react @formily/antd @formily/next 如果版本不一致也会重复打包

**Comment by tgxpuisb** [comment]
我这边遇到过一个案例，基于 `formily` 先封装一个组件 `xxxFormilyComponent` 然后组件发布npm包之后再引用到项目中，这种情况极容易造成重复打包问题，我的解决方案如下，参考了 `antd` 二次开发组件的打包思路

1. 将 `xxxFormilyComponent` 组件中 `package.json中` 所有的 `@formily/*` 的组件全部放入 "peerDependencies" 中
![image](https://user-images.githubusercontent.com/11400784/118074768-37327800-b3e1-11eb-89b3-3f73a3f7a818.png)

2. 利用 extenal 机制，在打包出来的代码中不再直接引入 `@formily/*` 中任何代码内容，**参考 `antd` 的二次封装组件库不会打包 `antd` 和 `react` 这一思路**

3. 在业务代码中 引入 `xxxFormilyComponent` ，同时也引入 `@formily/*` 的相关代码
![image](https://user-images.githubusercontent.com/11400784/118075077-e40cf500-b3e1-11eb-92f7-ccc7341c8689.png)

此时业务中就只会打包出一份 `@formily/*` 的代码了

**Comment by janryWang** [comment]
https://github.com/alibaba/formily/issues/1830 这个也是重复打包问题，原因是babel-import-plugin没配对，走了lib目录，但是formily内部是走的es
